what4/src/What4/Protocol/SMTWriter.hs

```diff
@@ -2035,6 +2036,8 @@ appSMTExpr ae = do
     RealDiv xe ye -> do
       x <- mkBaseExpr xe
       case ye of
+        SemiRingLiteral SR.SemiRingRealRepr r _ | r == 0 -> do
+          freshBoundTerm RealTypeMap $ rationalTerm infinity
```

Comment:
I don't think I'm really comfortable with this change.  What's the reasoning?

Reply:
It is a somewhat unsatisfactory change, I agree. It came up from an SV-COMP benchmark that involves verification of the `atan` and `atan2` functions on floats. In this code, there's a branch that does a division by a value that's initialized to zero. It may be (and may not be?) the case that that particular situation is unreachable. However, we still wind up building and sending to the SMT solver a division by zero. Without this case, the program just crashes with a `zero denominator` message from a call to `error`.

When working with integers, we've usually made division total and added a side condition to ensure that the denominator is non-zero. With reals, something similar is probably sensible. With floats, the "correct" result depends on what mode you're in. Infinity is one of the legal values, as is NaN.

Reply:
It looks like we're avoiding using the division operator for some reason.  SMTLib2 and Yices both support a division operator, so we don't actually need to do that.  We should treat this just the same way as integer division and use the division operator.

Reply:
Oh, yeah, we should just use the SMTLib division operator (making sure to avoid concrete division). I'll change that.

Reply:
On closer inspection, this is a little harder. I think the reason for the `x * rationalTerm (recip r)` is that the `SupportTermOps` class uses a `Num` superclass for numeric operations, which includes `*` but not `/` or `div`. And it's the `recip` operation (on a concrete value) that led to the error. The `SupportTermOps` class has named `intDiv` and `floatDiv` functions, but nothing for reals.

Reply:
OK, let's do this: the immediate fix is just to add a guard that `r /= 0` before calling `recip` and allow it to fall into the second case if it is.  Separately, let's make a ticket for adding direct support for real division.

Reply:
Turns out it was easy to add real division. See  28feb2e.